jetty-fcgi/fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/HttpChannelOverFCGI.java

```diff
@@ -45,104 +43,121 @@
     private static final Logger LOG = LoggerFactory.getLogger(HttpChannelOverFCGI.class);
 
     private final Queue<HttpInput.Content> _contentQueue = new LinkedList<>();
-    private final AutoLock _lock = new AutoLock();
-    private HttpInput.Content _specialContent;
+    private final Callback _asyncFillCallback = new AsyncFillCallback();
+    private final ServerFCGIConnection connection;
     private final HttpFields.Mutable fields = HttpFields.build();
     private final Dispatcher dispatcher;
+    private HttpInput.Content _specialContent;
     private String method;
     private String path;
     private String query;
     private String version;
     private HostPortHttpField hostPort;
 
-    public HttpChannelOverFCGI(Connector connector, HttpConfiguration configuration, EndPoint endPoint, HttpTransport transport)
+    public HttpChannelOverFCGI(ServerFCGIConnection connection, Connector connector, HttpConfiguration configuration, EndPoint endPoint, HttpTransport transport)
     {
         super(connector, configuration, endPoint, transport);
+        this.connection = connection;
         this.dispatcher = new Dispatcher(connector.getServer().getThreadPool(), this);
     }
 
     @Override
     public boolean onContent(HttpInput.Content content)
     {
-        boolean b = super.onContent(content);
+        boolean result = super.onContent(content);
 
-        Throwable failure;
-        try (AutoLock l = _lock.lock())
-        {
-            failure = _specialContent == null ? null : _specialContent.getError();
-            if (failure == null)
-                _contentQueue.offer(content);
-        }
+        Throwable failure = _specialContent == null ? null : _specialContent.getError();
+        if (failure == null)
+            _contentQueue.offer(content);
         if (failure != null)
             content.failed(failure);
 
-        return b;
+        return result;
     }
 
     @Override
     public boolean needContent()
     {
-        try (AutoLock l = _lock.lock())
+        if (needContent("immediate"))
+            return true;
+        parseAndFill();
+        if (needContent("parsed"))
+            return true;
+        connection.getEndPoint().tryFillInterested(_asyncFillCallback);
+        return false;
+    }
+
+    private boolean needContent(String context)
```

Comment:
I find this method name slightly confusing as it doesn't "demand" content like all other `needContent` methods do. How about calling it `hasContent`?

Reply:
Done.